### PR TITLE
[fix][client] Add initialization for the OpSendMsg

### PR DESCRIPTION
**Pull request by RobertIndie**
### Motivation

Some parameters in OpSendMsg are not specified with initialized values. 

These initial values are all relied upon by other codes. Such as here:
https://github.com/apache/pulsar/blob/eec46ddcba4d2b4f956e1b4d63154cc43087f507/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L1409-L1422

### Modifications

* Add the initialization for the OpSendMsg.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 1]
```diff
@@ -1337,15 +1337,15 @@ protected static final class OpSendMsg {
         SendCallback callback;
         Runnable rePopulate;
         ChunkedMessageCtx chunkedMessageCtx;
-        long uncompressedSize;
-        long sequenceId;
-        long createdAt;
-        long firstSentAt;
-        long lastSentAt;
-        int retryCount;
```
Comment (HQebupt) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:1345]: Does it have any effect if no explicit initial value ?

Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:1345]: Yes. These initial values are all relied upon by other codes. Such as here:
https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L1409-L1422

**Review by BewareMyPower — CHANGES_REQUESTED** [review]
Please explain what's the affect of these fields without initial values.

**Comment by RobertIndie** [comment]
> Please explain what's the affect of these fields without initial values.

I have explained it in the PR description. Is there something unclear?

**Comment by BewareMyPower** [comment]
In addition, adding initial values to a recyclable class is dangerous. For a recyclable class, we should make constructors private and only expose some factory methods and set the value in these factory methods.

The reason is that if the object was reused from the pool, the initial value should be what was set in `recycle()` method.

For example,

```java
@Data
class RecyclableInteger {
    private static final Recycler<RecyclableInteger> RECYCLER = new Recycler<RecyclableInteger>() {
        @Override
        protected RecyclableInteger newObject(Handle<RecyclableInteger> handle) {
            return new RecyclableInteger(handle);
        }
    };

    private final Recycler.Handle<RecyclableInteger> handle;
    private int x = 100; // initial value: 100

    private RecyclableInteger(Recycler.Handle<RecyclableInteger> handle) {
        this.handle = handle;
    }

    public static RecyclableInteger create() {
        return RECYCLER.get(); // retrieve an object without modifying the value of `x`
    }

    public void recycle() {
        this.x = 0; // recycled value: 0
        handle.recycle(this);
    }
}
```

```java
        RecyclableInteger i = RecyclableInteger.create();
        System.out.println(i.getX()); // 100 as we expected
        i.recycle();
        i = RecyclableInteger.create();
        System.out.println(i.getX()); // 0, which is unexpected
```

**Comment by BewareMyPower** [comment]
> I have explained it in the PR description. Is there something unclear?

Sorry at the time I replied, the PR description was not updated. Then, please see my another comment.

**Comment by BewareMyPower** [comment]
The best practice might be adding an `initialize()` method to a recyclable class and calling this method at the beginning of each factory method, like:

```java
    private void initialize() {
        x = 100;
    }

    public static RecyclableInteger create() {
        final RecyclableInteger i = RECYCLER.get();
        i.initialize();
        return i;
    }
```

**Comment by RobertIndie** [comment]
> In addition, adding initial values to a recyclable class is dangerous. For a recyclable class, we should make constructors private and only expose some factory methods and set the value in these factory methods.
> 
> The reason is that if the object was reused from the pool, the initial value should be what was set in `recycle()` method.
> 

Before this PR, there is already an inconsistent issue in the OpSendMsg. The value of the OpSendMsg just initialized is not the same as the OpSendMsg just being recycled. This PR is to fix this inconsistency. I was wondering if we could have the recyle method call some kind of initialization method so we don't need to maintain the same initialization value in two places.

**Comment by BewareMyPower** [comment]
> if we could have the recyle method call some kind of initialization method so we don't need to maintain the same initialization value in two places.

Yes.

**Comment by RobertIndie** [comment]
@BewareMyPower Hi, I have added the initialization for the OpSendMsg, PTAL again. Thanks.

**Review by BewareMyPower — APPROVED** [review]

**Review by nodece — APPROVED** [review]
LGTM

**Pull request merged**